mslaved/mslavedc.8
.TH MSLAVEDC 8 "" "MandelSpawn"
.SH NAME 
mslavedc \- MandelSpawn computation server control program
.SH SYNOPSIS  
.B mslavedc \-k 
|
.B \-q 
host | [
.B \-p 
port ] [ 
.B -n 
niceval ] [
.B \-t 
timeout ] mslaved
.SH DESCRIPTION  
.I mslavedc
is a program for starting and killing \fImslaved\fP processes on
machines where this is not done automatically by 
.IR inetd (8C).
.I Mslavedc 
is used by the 
.IR enslave (1)
script.
.PP
With the \fB\-q\fP option (query), it tries to contact a server
already running on the specified host.  If it succeeds, it prints the
PID of that server.  
.PP
With the \fB\-k\fP option, mslavedc tries to kill any \fImslaved\fP 
process already running on the local host.
.PP
If neither \fB\-k\fP nor \fB\-q\fP is given, \fImslavedc\fP 
expects the pathname of 
.IR mslaved (8)
as an argument.  A datagram socket is created and 
\fImslaved\fP is \fIexec\fP'ed with the socket as standard input. The
\fB\-p\fP option may be used to specify a port number other than the
default of 9359.  The \fB\-n\fP option and its argument, if present,
are passed on to \fImslaved\fP to set its nice value.
The \fB\-t\fP (timeout) option is also passed to
\fImslaved\fP; if not present, \fB\-t900\fP is used giving a timeout of
15 minutes.
.PP
.SH "SEE ALSO"
xms(1), bms(1), enslave(1), mslavedc(8)
.PP
.SH COPYRIGHT
.if n Copyright (C) 1990 Andreas Gustafsson (gson@niksula.hut.fi)
.if t Copyright \(co 1990 Andreas Gustafsson (gson@niksula.hut.fi)


ChangeLog
Wed Aug  6 17:23:33 1997  Andreas Gustafsson  <gson@guava.araneus.fi>

	* Wrapped version 0.07pl2

	* Switched from compress to gzip.

	* Updated contact information in README.

	* Commented out the X11R[23] support in xms/Imakefile
	because it misidentified X11R6 as X11R[23], and no-one
	uses X11R[23] anymore anyway.

	* 0.07pl1 fixed select() bug that slowed down bms to one packet
        per second.

Sat Sep  4 03:08:49 1993  Andreas Gustafsson  (gson@guano)

	* Wrapped version 0.07

	* Fixed two bugs in Ms.c which caused core dumps under
	Linux

Sun Aug 15 12:11:38 1993  Andreas Gustafsson  (gson@guano)

	* Added Linux FNDELAY hack
	
	* Added strdup() replacement (needed by pure 4.3BSD)

Sun Jun  6 04:27:00 1993  Andreas Gustafsson  (gson@guano)

	* Rewrote parts of the README file.

	* Gave up futile attempts to declare malloc() using system 
	include files.

	* Combined the various INSTALL files into a single file.

	* Fixed an use of freed memory and several memory leaks.

Sat May 22 03:04:12 1993  Andreas Gustafsson  (gson@guano)

	* Updated years in copyright messages.

Fri May  7 04:41:48 1993  Andreas Gustafsson  (gson@guano)

	* Version 0.06g ported to HP700-series and DEC Alpha
	machines.  The rubberband box can now be manipulated
	while the calculation is in progress.

	* Version 0.06h was put under RCS.  This necessitated a 
	restructuring of the code into a more traditional 
	directory tree with no links between directories.  There
	is now a lib/ directory for code shared between the
	different programs.  PPM support was added to bms.

	* Version 0.06f introduced experimental Henon fractal code
	in mslaved and associated generic parameter decoding code.

	* Version 0.06d1 changed to support a BitMapUnit of 32 in
	b/w operation.  Maybe it was a mistake to remove the
	8-bit code.

Thu May 14 19:58:16 1992  Andreas Gustafsson  (gson@psykoman.cs.hut.fi)

	* Lots of changes to support 24-bit visuals

Sat Feb  8 21:15:44 1992  Andreas Gustafsson  (gson at gunnar)

	* Added support for pipe operation in mslaved.c; introduced
	HAVE_SOCKETS and HAVE_SYSLOG compile-time flags

Wed Feb  5 16:57:41 1992  Andreas Gustafsson  (gson at spiff.cs.hut.fi)

	* Made the visual matching code	use the default visual for 
	black-and-white operation

	* Added colormap animation

Fri Dec  6 17:32:39 1991  Andreas Gustafsson  (gson at bulldozer.cs.hut.fi)

	* Allowed empty lines and # comments in .enslave

Sun Dec  1 16:18:26 1991  Andreas Gustafsson  (gson at punisher.cs.hut.fi)

	* New development version 0.06a.  Includes visual matching code in
	Mama.c, and new option handling in mslaved and mslavedc.

Sun Nov 24 12:38:12 1991  Andreas Gustafsson  (gson at hydroman.cs.hut.fi)

	* Wrapped version 0.06

	* Added option details to the synopsis in xms.man

	* Tested the new ms_draw() on monochrome and 8-bit colour
	Suns.

	* Completely rewrote the black-and-white special case display
	support in Ms.c:ms_draw(), so that it now constructs the image
	a byte at a time rather than a word at a time.  This should
	avoid any byteorder problems.

Tue Nov  5 17:23:38 1991  Andreas Gustafsson  (gson at bulldozer.cs.hut.fi)

	* Version number is now 0.05p.  Changed some key bindings to be 
	more consistent with the new "zoom by default" policy, and
	updated the manual page correspondingly.

Mon Sep 30 01:01:40 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* Added XtNinput argument to shell widget; that ought to
	fix the keyboard focus problem with OpenWindows (not tested).

Sun Sep 22 12:07:27 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* Edited README and mslaved/Makefile; now gcc is stronly recommended
	only for old CISCs.  Made floating point the default for MIPS in
	ms_real.h as suggested by broadley's benchmarks.

	* restored the menu entries for using the old window but made the
	popup alternative more defaultish (no, I don't run spell on
	ChangeLogs). 

	* added #include <sys/file.h> to ms_ipc.h (needed for FNDELAY in
	HP-UX) 

	* fixed some syntax errors in mslaved.c

	* added #include <netinet/in.h> in Ms.c to get htonl() and ntohs()
	macro definitions (needed for HP-UX)

Sat Sep 14 00:09:20 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* #ifdef'ed away the menu entries for reusing the old window

Sun Aug 11 14:42:55 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* Added "inverse box zooming" as suggested by RMS.  Still
	untested.

Sat Aug 10 10:32:51 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* Added #include <errno.h> in work.c

	* Wrapped 0.05j and mailed to broadley for beta testing

	* Renamed split.c to mspawn.c

	* Tested 0.05h on SPARCstation IPC.  Fixed some bugs in the
	noImakefile and documented problems with missing imake and
	Aw libs.  Version with these fixes called 0.05i.

Fri Jul 26 14:23:57 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* Fixed some more typos and made some other changes in the man
	pages

	* Wrapped 0.05h	and gave it to ufa@niksula.hut.fi for beta 
	testing 

Sun Jul 21 12:36:22 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* Wrote the bms batch mode client and renamed the X client to
	xms.  This also caused lots of changes to both source and 
	documentation of xms because of extensive code sharing between the
	two clients.  Mandelbrot-dependent but X-independent
	code has been separated into a new module, mspawn.c, leaving
	code that is both X- and Mandelbrot-independent in work.c.
	The timeout system has been completely rewritten.

	* MIPS fixed-point code has been tested on a DECStation.

Wed Jun 26 17:01:04 1991  Andreas Gustafsson  (gson at gunnar)

	* Separated the X-independent workload distribution stuff
	into a new module, work.c.  Renamed all X-independent files
	to lowercase names to reduce ugliness.  Added missing "2" in
	MIPS fracmult2 function name; need to retest MIPS code before
	distributing it.

Tue Jun 18 21:34:57 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* Fixed several typos in the man page


mslaved/mslavedc.c
/* mslavedc.c - MandelSpawn computation server control program */
/* (named in accordance with timedc, etc.). */

/*  
    This file is part of MandelSpawn, a network Mandelbrot program.

    Copyright (C) 1990-1993 Andreas Gustafsson

    MandelSpawn is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License, version 1,
    as published by the Free Software Foundation.

    MandelSpawn is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License,
    version 1, along with this program; if not, write to the Free 
    Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include <stdio.h>
#include <errno.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <signal.h>

/*
  The only system I know needs the following two #includes is SCO Unix, 
  which needs them to define FNDELAY.  If they cause trouble with your
  system, just remove them.
*/
#include <sys/types.h>
#include <sys/file.h>

/* This is for Linux */
#ifndef FNDELAY
#ifdef O_NDELAY
#define FNDELAY O_NDELAY
#endif
#endif

#include "datarep.h"
#include "ms_ipc.h"

#define RETRIES 5		/* how many times to retry getting a PID in ping() */

extern int errno;		/* 4.3BSD needs this even when errno.h is included */

char *me;			/* name of program */
int port = DEFAULT_PORT;	/* UDP port */
char *command = NULL;

int kill_mode = 0;

int timeout_supplied = 0;	/* true if user gave a timeout option */

/* Report a fatal error and exit */

void error(s)
char *s;
{
	fprintf(stderr, "%s: ", me);
	perror(s);
	exit(1);
}

/*
  Repeatedly send datagrams to "host" trying to find out the pid of 
  an existing server, if any.  Returns the pid, or 0 if no response.
*/

int ping(host)
char *host;
{
	int i;
	int status;
	WhoAreYouMessage out;
	Message in;
	int osock, isock;
	struct sockaddr_in oname, iname, mname;
	int mnamelen;
	struct hostent *hp, *gethostbyname();
	/* set up the output socket; use an arbitrary port */
	osock = socket(AF_INET, SOCK_DGRAM, 0);
	if (osock < 0) {
		error("opening output socket");
	}
	hp = gethostbyname(host);
	if (hp == 0) {
		error("unknown host");
	}
	/* set up the destination address */
	oname.sin_family = AF_INET;
	oname.sin_port = htons(port);
	bcopy(hp->h_addr, (char *)&oname.sin_addr, hp->h_length);
	isock = socket(AF_INET, SOCK_DGRAM, 0);
	if (isock < 0) {
		error("opening input socket");
	}
	iname.sin_family = AF_INET;
	iname.sin_addr.s_addr = INADDR_ANY;
	iname.sin_port = 0;
	if (bind(isock, &iname, (int)sizeof(iname))) {
		error("binding input socket");
	}
	if (fcntl(isock, F_SETFL, FNDELAY) == -1) {
		error("unblocking socket");
	}
	mnamelen = sizeof(mname);
	if (getsockname(isock, &mname, &mnamelen) == -1)
		error("getsockname");
	out.header.magic = htons(MAGIC);
	out.header.version = htons(VERSION);
	out.header.type = htons(WHO_R_U_MESSAGE);
	out.port = mname.sin_port;	/* in network byte order already */
	for (i = 0; i < RETRIES; i++) {
		int nbytes;
		int fromlen;
		if (sendto(osock, (char *)&out, sizeof(WhoAreYouMessage), 0,
			   (struct sockaddr *)&oname, sizeof(oname)) < 0)
			error("sending pid inquiry message");
		/*
		   This previously used read(), but some non-BSD TCP/IP implementations
		   don't allow it to be used with connectionless sockets.  Also,
		   while the Sun implementation does allow for a null pointer
		   for the "from" argument in recvfrom(), the "fromlen" argument
		   may not be a null pointer.
		 */
		nbytes = recvfrom(isock, (char *)&in, sizeof(in),
				  0, (struct sockaddr *)0, &fromlen);
		if (nbytes == 0
#ifdef EWOULDBLOCK
		    || (nbytes == -1 && errno == EWOULDBLOCK)
#endif
		    ) {		/* no reply yet */
			sleep(1);
		} else if (nbytes == -1) {
			error("receiving pid info");
		} else {	/* got a reply */
			status = ntohs(in.iam.pid);
			goto done;
		}
	}
	status = 0;		/* unsuccessful */
 done:
	(void)close(osock);
	(void)close(isock);
	return (status);
}

/* 
  Kill any server that is already active in this machine.
  Return 0 if successful kill, 1 if no response.
*/

int murder()
{
	int pid;
	char my_name[256];
	if (gethostname(my_name, sizeof(my_name)) == 1)
		error("gethostname");
	pid = ping(my_name);	/* get pid of existing server, if any */
	if (pid == 0)
		return (1);	/* return failed exit status */
	if (kill(pid, SIGTERM) == -1)
		error("kill");
	return (0);
}

/* argument list for mslaved */
static char *arglist[16];	/* more than enough */
static char **arglistp = arglist + 1;	/* pointer to the above */

void copy_arg(p)
char *p;
{
	*arglistp++ = p;
}

/* Set up the file handles and exec an mslaved */

void birth()
{
	int isock;
	struct sockaddr_in iname;
	int tty;
	int i;

	/* set up the socket */
	isock = socket(AF_INET, SOCK_DGRAM, 0);
	if (isock < 0) {
		error("opening socket");
	}
	iname.sin_family = AF_INET;
	iname.sin_addr.s_addr = INADDR_ANY;
	iname.sin_port = htons(port);
	if (bind(isock, &iname, (int)sizeof(iname)) == -1) {	/* probable cause of error is that the port is in use */
		error("binding socket");
	}

	/* dup the newly-opened socket to stdin */
	close(0);
	dup(isock);

#ifdef TIOCNOTTY		/* presumably BSD-like */
	if ((tty = open("/dev/tty", 0, 0)) != -1) {
		ioctl(tty, TIOCNOTTY, (struct sgttyb *)0);
		(void)close(tty);
	} else {
		if (setpgrp(0, 0))
			error("setpgrp");
	}
#else				/* presumably SYSV-like */
	setpgrp();
#endif

	for (i = 1; i < 10; i++)
		(void)close(i);

	arglist[0] = command;

	/* supply a default timeout if none was given */
	if (!timeout_supplied)
		copy_arg("-t900");

	copy_arg((char *)0);
	execvp(command, arglist);
	exit(1);		/* there's no good way to report errors here */
}

int main(argc, argv)
int argc;
char **argv;
{
	char *s;

	me = argv[0];
	for (; s = *++argv, --argc;) {
		if (*s != '-') {
			command = s;
		} else {
			switch (s[1]) {
			case 'k':	/* kill */
				kill_mode++;
				break;
			case 'q':	/* query */
				{
					int rpid =
					    ping(s[2] ? s +
						 2 : (--argc, *++argv));
					if (rpid) {
						printf("%d\n", rpid);
						exit(0);
					} else
						exit(1);
				}
			case 't':	/* timeout */
				timeout_supplied = 1;
				/* fallthrough */
			case 'n':	/* nice */
				/* copy these options and possible separate option arguments to */
				/* the mslaved argument list */
				copy_arg(s);
				if (!s[2])
					copy_arg((--argc, *++argv));
				break;
			case 'p':	/* port */
				port = atoi(s[2] ? s + 2 : (--argc, *++argv));
				break;
			default:
				goto usage;
			}
		}
	}

	if (kill_mode)
		exit(murder());

	if (!command)
		error("missing command");

	switch (fork()) {
	case -1:
		error("fork");
	case 0:		/* child */
		birth();
	default:		/* parent */
		exit(0);
	}
 usage:
	fprintf(stderr, "usage: %s [-k] [-q] [-nnice] [-ttimeout] [-pport]\n",
		me);
	exit(1);
}
